cni/pkg/nodeagent/server.go

```diff
@@ -0,0 +1,249 @@
+		RedirectDNS:   true, // TODO Currently for ambient, DNS redirection must happen to properly resolve serviceentries.
+	}
+
+	iptablesConfigurator := iptables.NewIptablesConfigurator(cfg, realDependencies(), iptables.RealNlDeps())
+
+	// Create hostprobe rules now, in the host netns
+	// Later we will reuse this same configurator inside the pod netns for adding other rules
+	iptablesConfigurator.DeleteHostRules()
+
+	if err := iptablesConfigurator.CreateHostRulesForHealthChecks(&HostProbeSNATIP); err != nil {
+		return nil, fmt.Errorf("error initializing the host rules for health checks: %w", err)
+	}
+
+	podNetns := NewPodNetnsProcFinder(os.DirFS(filepath.Join(pconstants.HostMountsPath, "proc")))
+	netServer := newNetServer(ztunnelServer, podNsMap, iptablesConfigurator, podNetns, set)
+
+	// Set some defaults
+	s := &Server{
+		ctx:        ctx,
+		kubeClient: client,
+		isReady:    ready,
+		dataplane: &meshDataplane{
+			kubeClient: client.Kube(),
+			netServer:  netServer,
+		},
+	}
+	s.NotReady()
+	s.handlers = setupHandlers(s.ctx, s.kubeClient, s.dataplane, args.SystemNamespace)
+
+	cniServer := startCniPluginServer(ctx, pluginSocket, s.handlers, s.dataplane)
+	err = cniServer.Start()
+	if err != nil {
+		return nil, fmt.Errorf("error starting cni server: %w", err)
+	}
+	s.cniServerStopFunc = cniServer.Stop
+
+	return s, nil
+}
+
+func (s *Server) Ready() {
+	s.isReady.Store(true)
+}
+
+func (s *Server) NotReady() {
+	s.isReady.Store(false)
+}
+
+// buildKubeClient creates the kube client
+func buildKubeClient(kubeConfig string) (kube.Client, error) {
+	// Used by validation
+	kubeRestConfig, err := kube.DefaultRestConfig(kubeConfig, "", func(config *rest.Config) {
+		config.QPS = 80
+		config.Burst = 160
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed creating kube config: %v", err)
+	}
+
+	client, err := kube.NewClient(kube.NewClientConfigForRestConfig(kubeRestConfig), "")
+	if err != nil {
+		return nil, fmt.Errorf("failed creating kube client: %v", err)
+	}
+
+	return client, nil
+}
+
+// createHostsideProbeIpset creates an ipset. This is designed to be called from the host netns.
+// Note that if the ipset already exist by name, Create will not return an error.
+//
+// We will unconditionally flush our set before use here, so it shouldn't matter.
+func createHostsideProbeIpset() (ipset.IPSet, error) {
+	linDeps := ipset.RealNlDeps()
+	probeSet, err := ipset.NewIPSet(constants.ProbeIPSet, linDeps)
+	if err != nil {
+		return probeSet, err
+	}
+	probeSet.Flush()
+	return probeSet, nil
+}
+
+func (s *Server) Start() {
+	log.Info("CNI ambient server starting")
+	s.kubeClient.RunAndWait(s.ctx.Done())
+	log.Info("CNI ambient server kubeclient started")
+	pods := s.handlers.GetAmbientPods()
+	err := s.dataplane.ConstructInitialSnapshot(pods)
+	if err != nil {
+		log.Warnf("failed to construct initial snapshot: %v", err)
+	}
+
+	log.Info("CNI ambient server marking ready")
+	s.Ready()
+	s.dataplane.Start(s.ctx)
+	s.handlers.Start()
```

Comment:
> A more typical flow would be to remove the ConstructInitialSnapshot and wait for the handler to process all initial events.

How would we know what the "initial" events are?

Reply:
> How would we know what the "initial" events are?

You just process all events, special handling of "initial" seems likely to be broken

> Pod 3 wouldn't be skipped by the informer, as it would not have been annotated by either the plugin or the informer yet.

`case controllers.EventAdd:` - does nothing, no?

Reply:
> `case controllers.EventAdd:` - does nothing, no?
It falls thru to `controllers.EventUpdate`, and should handle the add event - in that block, if the pod should be enrolled, but does not have an annotation indicating it has already been enrolled, it will be enrolled.

The plugin, and the controller can both add that annotation, depending on what catches it first. As long as it has been caught by $something, in that block, it will be skipped. If it has not, it will be enrolled.

Reply:
Go switch statements do not fallthrough (unless you explicitly `fallthrough`).

Reply:
derp, I'm confusing my languages.

Yes, ambient pod enrollment before the CNI server is in a _ready_ state can have edge case behavior - that's part of what https://github.com/istio/istio/issues/48286 is intended to address.

A way around this would be to not label anything for ambient inclusion before the CNI installed. This also creates issues for upgrades, but CNI upgrades as as-yet-unsolved.

Reply:
Raised https://github.com/istio/istio/issues/49009 since we probably need a CNI upgrade story - I think simply making sure we don't "miss" pods during upgrade is enough, which might involve something like the sidecar detect+repair, but I'm not sure.